Feature add data logging

AdamBusch opened the pull request:
This feature will log every can frame to a text file for later analysis in Matlab. Formatted to match the log files produced by wifican.js

AdamBusch committed:
Add functionality for reading serial data

Serial data is read from the CANdapter and sent to all clients connected
on websockets.
Improved client view shows data in realtime graph.

AdamBusch committed:
Fixed linting issues

Single line was too long...

AdamBusch committed:
Import full .gitignore file

Fixes #1

AdamBusch committed:
Add README.md

Our repo doesn't have a README, this commit adds one.

AdamBusch committed:
Merge branch 'feature_add_serial_reading' of https://github.com/PurdueElectricRacing/Wireless-Telemetry-Server into feature_add_serial_reading

AdamBusch committed:
Add inital data logging

Output data logs to the ./logs directory and mimic the styel of past
logs made with the wifican.js file to allow for legacy Matlab
processing.

AdamBusch committed:
Add final data formatting and IP address

Add final ip configuration for use with router on car

AdamBusch committed:
Remove clientside code

This code is unessicary, replaced by KX dashboard

AdamBusch committed:
Add router configuration files

Add backup of router configuration. Can be loaded by the router control
panel.

AdamBusch committed:
Fix linting issues.

jacklecroy committed:
Added a period

AdamBusch committed:
Merge branch 'master' into feature_add_data_logging

RyanSchw commented:
When you pull `master` into this, you may get merge conflicts with `.gitignore`. Make sure you include `logs/` when you resolve conflicts.

RyanSchw commented on DAQ_Simulation.py:
```diff
@@ -10,21 +10,50 @@
 
 
 async def run_server(ip, port):
+    print('Hello?')
```
Remove debugging comment

RyanSchw commented on DAQ_Simulation.py:
```diff
@@ -10,21 +10,50 @@
+    print('Hello?')
     wait_task = asyncio.ensure_future(send_data())
     server_task = asyncio.ensure_future(server.get_server('127.0.0.1', 5000))
-    print(f'Serving websocket server on {ip}:{port} ...')
```
Is the `ip` and `port` documented somewhere else so people new to the project know where to find this information?

AdamBusch replied:
Yes, it is located on the wiki [here](http://purdueelectricracing.com/wiki/index.php/Wireless_Telemetry)

RyanSchw commented on Main_WT_Server_DAQ.py:
```diff
@@ -0,0 +1,128 @@
+multi_frame_message = ''
+
+
+# # Logs a set of CAN frames in a time interval to a log file
```
Double comment

AdamBusch committed:
Fix debug messages and double comment

Remove debugging print and double comment to make ryan happy so he can
merge this old pull request. This is important because this code is 3
months old and needs to be put on master eventually.

RyanSchw commented:
@AdamBusch Did I merge out of order? Just did #6, so you might need to pull from `master` again. Gonna have to merge conflicts when you do so.

bsmrdel commented on DAQ_Simulation.py:
```diff
@@ -10,21 +10,49 @@
+        'id': '501',
+        'data': '0F900EBF02A0029F'
+    },
+]
```
Would it be wise to include which CAN ID each of these are? For example, 421 is IMU data.

bsmrdel replied:
Although I guess it is documented [here](https://docs.google.com/spreadsheets/d/195Y2cf9C2mrA6QCLuLJv3NHsCqTLi-3Wd07q4T5rXuM/edit#gid=0). Maybe we just need to clean that up instead.

RyanSchw replied:
Perhaps we could leave a comment at the top explaining where to find the CAN IDs?

AdamBusch replied:
These are just for simulating the DAQ data coming from the CAN bus, should we still try to document this here? I feel like it would be better documented somewhere else when we actually work on parsing the data.

RyanSchw replied:
The point of the comment would be to better explain what data is being passed. When writing test code (or in this case, simulation code) I think it's always best to explain what data is being provided in order to give some background. But that's just me

RyanSchw reviewed (dismissed):
`master` should be pulled into this before we review it again. When master is pulled in, it will also mean @AdamBusch will have to resolve the merge conflicts (which should be done now before merging).

AdamBusch committed:
Merge remote-tracking branch 'origin/master' into feature_add_data_logging

AdamBusch committed:
Add initial sensor dictionaly with CAN IDs

This is a somewhat complete list of snesors on the DCAN bus that will be
sent over the wireless telemetry server. We need to add additional
sensors on the DCAN such as BMS info and finalize the IDs for each
sensor.

AdamBusch committed:
Add potential IDs and byte indicies.

First revision of CAN information for new sensors

AdamBusch committed:
Update message ID formatting

AdamBusch committed:
Remove client HTML code

Static HTML client is no longer needed for viewing data

AdamBusch committed:
Update tire temp byte length

Ture temp is being sent as a double value, needs 8 bits to represent
it.

jacklecroy committed:
Adds a function to read sensor data from dictionary

data_to_dict Takes in data and reformats it to make it more readable. As long as the SensorDict.csv stays in the same format, this probably won't need many changes in the future.

AdamBusch committed:
Add print formatting for excel/google sheets

Adds an option for a console printout of the CSV file to be pasted into
the excel sheet that keeps track of all the CAN IDs

jacklecroy committed:
Adds function to save sensor data from a list

Also updated data_to_dict to include "timestamp" in its output. This is so that these functions can be used together, both using timestamps.

jacklecroy committed:
Merge branch 'feature_add_sensor_translation' of https://github.com/PurdueElectricRacing/Wireless-Telemetry-Server into feature_add_sensor_translation

AdamBusch committed:
Output correct start/stop bits for can IDs

jacklecroy committed:
Update SensorDict.py to coding standards

Used autopep8 to update the file.

jacklecroy committed:
Updates SensorDict.py with line limit 80

Limit was not 80 when I ran autopep before.

AdamBusch committed:
Merge branch 'feature_add_sensor_translation' of https://github.com/PurdueElectricRacing/Wireless-Telemetry-Server into feature_add_sensor_translation

AdamBusch committed:
Change CAN message format for IMU

AdamBusch committed:
Fix linting issues

Travis plz

AdamBusch committed:
Add support for parsing multiple data points

Will import data from 'SensorData.csv' and parse incoming data packets
based on the information peovided in the file. This file is used to
define each sensor's address and start/end bytes used across the CAN bus
and wireless Telemetry server.

AdamBusch committed:
Split sensors to front/back of car

Sensors on the same CAN ID cannot be sent from oppisite sides of the
car, need to be grouped into which MC is sending their data (front/back)

Encapsulate CSV reading into class

This will allow for us to pre-load the sensor dict to speed up parsing
individual data frames.

Re-arrange dict format

MSB should be places before LSB as this is how it is formatted in the
CAN messages.

Add sample usage code and doccumentation

Sample code can be used as a quick unit test and to ensure that the CSV
format is valid

AdamBusch committed:
Merge branch 'feature_add_sensor_translation' of https://github.com/PurdueElectricRacing/Wireless-Telemetry-Server into feature_add_sensor_translation

AdamBusch committed:
Fix linting issues

RyanSchw commented on Main_WT_Server_DAQ.py:
```diff
@@ -3,17 +3,61 @@
 
 # Using different threads, this file will setup a
-# websocket server @ 192.168.4.1:5000 and broadcast
+# websocket server @ 192.168.4.1:5000 and broadcast.
```
It's fine for now, but at some point this should be removed. For a small typo like this, it's fine to not have a whole commit for this (just include it in some other feature)

RyanSchw commented on Main_WT_Server_DAQ.py:
```diff
@@ -3,17 +3,61 @@
+def log_CAN_data(timestamp, can_id, length, message):
+    current_time = int(round(time.time() * 1000))
+
+    # Wait for serial connection to stabalize, avoid garbage data
```
Stabilize, but like I said not that big of a deal.

RyanSchw commented on Main_WT_Server_DAQ.py:
```diff
@@ -24,6 +68,12 @@
 async def rec_serial_data(serial_connection):
+    current_time = time.time()
+    last_time = current_time
+    # buffer_time = 1/500
```
What's this?

AdamBusch replied:
This is used to keep track of the buffer time to ensure that we send data at a specified rate. Used to calculate delta time between loops.

RyanSchw reviewed (dismissed):
A few things:

1. Why is `PER-Rocket-*.cfg` and `PER-Bullet-*.cfg` named based on the datetime? Was this a specification of the config files? The date at which they're most up-to-date is tracked by GitHub (along with who made the changes, etc.) so there's no need (unless in the specs) to have the date in the file name.
2. There were a few typos, but it's not that big of a deal. Might just be a good idea to fix them in #8

AdamBusch commented:
> A few things:
> 
> 1. Why is `PER-Rocket-*.cfg` and `PER-Bullet-*.cfg` named based on the datetime? Was this a specification of the config files? The date at which they're most up-to-date is tracked by GitHub (along with who made the changes, etc.) so there's no need (unless in the specs) to have the date in the file name.
> 2. There were a few typos, but it's not that big of a deal. Might just be a good idea to fix them in #8

1. This was just how the config file was generated by the Ubiquity software, I figured it was best to not change it.

AdamBusch committed:
Fix typos mentioned in #7

RyanSchw commented:
> > A few things:
> > 
> > 1. Why is `PER-Rocket-*.cfg` and `PER-Bullet-*.cfg` named based on the datetime? Was this a specification of the config files? The date at which they're most up-to-date is tracked by GitHub (along with who made the changes, etc.) so there's no need (unless in the specs) to have the date in the file name.
> > 2. There were a few typos, but it's not that big of a deal. Might just be a good idea to fix them in #8
> 
> 1. This was just how the config file was generated by the Ubiquity software, I figured it was best to not change it.

OK, probably for the best then. May just be something to look into in the future.

Commit:
Merge pull request #8 from PurdueElectricRacing/feature_add_sensor_translation

Add initial sensor dictionary with CAN IDs

jacklecroy commented:
Honestly, I think you should squash the pull request when you merge it

AdamBusch committed:
Add C-compatiable names

We need additional names for generating the C header files.

jacklecroy committed:
Script to upate SensorDict from Google Sheets

Requires pandas and requests.
Saves the sensor info in the same format as SensorDict.py in file `output_file`
Ignores the first `ignore_rows` rows

jacklecroy committed:
Merge branch 'feature_add_data_logging' of https://github.com/PurdueElectricRacing/Wireless-Telemetry-Server into feature_add_data_logging

martinjaa18 committed:
update-SensorDict script adds comments

The comments column in the
spreadsheet will be used. The first byte
of each sensor will have a comment if
applicable.

jacklecroy reviewed (commented):
This was my commit, I don't know why it has this name, though.

RyanSchw commented:
Before this PR gets too large, is this at a point where it can be reviewed/merged? I can review it on Friday this week.

bsmrdel commented on SavedGoogleDrive.csv:
```diff
@@ -0,0 +1,53 @@
+0x740,0,1,ARBDropLink_Left,
+0x740,2,3,ARBDropLink_Right,
+0x740,4,5,FL ARB Droplink,
+0x740,6,7,FR ARB Droplink,
```
Why are the droplinks in here twice (lines 19-22)? Or do they require multiple bytes? Should either be removed if duplicate or named the same thing. Or I'm misunderstanding what's going on here.

jacklecroy replied:
This was automatically generated by a script based on what is in here: https://docs.google.com/spreadsheets/d/195Y2cf9C2mrA6QCLuLJv3NHsCqTLi-3Wd07q4T5rXuM/edit#gid=64209652

We can change the spreadsheet and it will reflect once you run the script again.

jacklecroy committed:
Add file to generate C headers

This script will read a SensorDict file
and generate C style headers for the
names and bytes.

jacklecroy committed:
Merge branch 'feature_add_data_logging' of https://github.com/PurdueElectricRacing/Wireless-Telemetry-Server into feature_add_data_logging

AdamBusch committed:
Format CANID names to work with header file gneration tool

RyanSchw commented on update-SensorDict.py:
```diff
@@ -0,0 +1,71 @@
+
+# Drop unwanted columns:
+content_df = content_df.drop(["TRANSMITTING MODULE", "DESCRIPTION / SENSOR",
+                              "AddMethd"], axis="columns")
```
Is this supposed to be `AddMethd`?

AdamBusch replied:
Yes, it is how it is formatted in the sheet.

AdamBusch committed:
Fix formatting errors and duplicate message ids.

AdamBusch commented:
We want to end development of this branch here. The features being added have gone way beyond adding data logging to master.

bsmrdel reviewed (approved):
Looks better now. Good by me, let's get this dude approved.

jacklecroy commented on GenerateHeaders.py:
```diff
@@ -0,0 +1,24 @@
+    ID = sensors_df.loc[row, "id"]
+    if lastID != ID:
+        savefile.write("#define " + sensor_name + "_CAN_ID " + ID + "\n")
+    savefile.write("#define " + sensor_name_and_type + "_START_BYTE " +
```
Good idea, didn't think of that

AdamBusch committed:
Fix linting errors...
